### PR TITLE
[Torch-to-Linalg] Add dynamic dimension support for BroadcastTo op

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit adds support for dynamic dimension in BroadcastTo op.

Signed-off-by: Abhishek Varma <abhishek@nod-labs.com>

### CODE REVIEW

**Comment by Abhishek-Varma** [comment]
Hi @ramiro050 could you please review this PR when you get a chance?

File: lib/Conversion/TorchToLinalg/Utils.cpp [hunk 1]
```diff
@@ -335,14 +336,18 @@ LogicalResult torch_to_linalg::broadcastToGivenShape(
...
   // Create affine map and shapes for tensor initialization.
   SmallVector<AffineExpr> outExpr;
   Value zero =
       rewriter.create<arith::ConstantOp>(loc, rewriter.getI64IntegerAttr(0));
+  Value zeroIndex =
+      rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(0));
+  Value oneIndex =
+      rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(1));
   size_t diff = broadcastToShape.size() - inputShape.size();
+  bool shiftToBroadcastShapeEmpty = shiftToBroadcastShape.size() == 0;
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Utils.cpp:350]: nit: unnecessary variable. You can use `shiftToBroadcastShape.empty()`

File: python/torch_mlir_e2e_test/test_suite/basic.py [hunk 2]
```diff
@@ -1333,6 +1333,33 @@ def BroadcastZeroRankInputStaticModule_basic(module, tu: TestUtils):
...
+    @annotate_args([
+        None,
+        ([1, -1, 1, -1], torch.float32, True),
+        ([1, -1, 1, -1], torch.float32, True),
+    ])
+    def forward(self, x, y):
+        dim_at_index_1 = torch.ops.aten.size(x, 1)
+        dim_at_index_3 = torch.ops.aten.size(x, 3)
+        res = torch.broadcast_to(y, [1, dim_at_index_1, 1, dim_at_index_3])
+        # print(res)
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/basic.py:1351]: remove print statement

File: python/torch_mlir_e2e_test/test_suite/basic.py [hunk 3]
```diff
@@ -1333,6 +1333,33 @@ def BroadcastZeroRankInputStaticModule_basic(module, tu: TestUtils):
...
+    @export
+    @annotate_args([
+        None,
+        ([1, -1, 1, -1], torch.float32, True),
+        ([1, -1, 1, -1], torch.float32, True),
+    ])
+    def forward(self, x, y):
+        dim_at_index_1 = torch.ops.aten.size(x, 1)
+        dim_at_index_3 = torch.ops.aten.size(x, 3)
+        res = torch.broadcast_to(y, [1, dim_at_index_1, 1, dim_at_index_3])
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/basic.py:1350]: use `torch.ops.aten.broadcast_to`

File: lib/Conversion/TorchToLinalg/DataMovement.cpp [hunk 4]
```diff
@@ -1179,12 +1179,26 @@ class ConvertAtenBroadcastToOp : public OpConversionPattern<AtenBroadcastToOp> {
       return rewriter.notifyMatchFailure(
           op, "unimplemented: the size list is not from list construct");
     }
+    SmallVector<bool> shiftToBroadcastShape;
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/DataMovement.cpp:1182]: I'm not sure I understand what this represents. Can you explain to me the reasoning?

Comment (Abhishek-Varma) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp]: Hi @ramiro050 ,

1. I've now renamed this to `useBroadcastToShape`.
2. `broadcastToShape` is already a construct which is/was previously being used in the lowering.
3. Now with `useBroadcastToShape` I'm trying to use `broadcastToShape` only for those dimensions which are dynamic and whose final shape needs to be inferred based on the "shape being broadcastedTo".

I've added a comment too in the file to address this.

Let me know if that sounds good to you.

**Comment by Abhishek-Varma** [comment]
Hi @ramiro050 

I've addressed your comments so far.
Can you take a look at the PR and re-review?

Also, the git checkout of submodules look like an orthogonal bug? @powderluv

**Comment by powderluv** [comment]
@ramiro050 is on vacation so please get someone else to review 

I will fix the CI

**Review by vivekkhandelwal1 — APPROVED** [review]
LGTM

**Comment by Abhishek-Varma** [comment]
Hi @powderluv 

Can we get this in and address any review comments (if any) later in a subsequent patch?

**Pull request merged**